port/linux/omrvmem.c

```diff
@@ -1174,13 +1180,33 @@ reserve_memory_with_mmap(struct OMRPortLibrary *portLibrary, void *address, uint
 			useBackingSharedFile = FALSE;
 			Trc_PRT_vmem_reserve_doubleMapAPI_not_available(address, byteAmount);
 		}
-	} else if (useBackingFile) {
+	} else if (useBackingSharedTmpFile) {
+		/* Generate a unique temporary filename from template and open the file. */
+		char filename[FILE_NAME_SIZE + 1];
+		snprintf(filename, sizeof(filename), "omrvmem_%09d_XXXXXX", getpid());
+		fd = mkostemp(filename, 0);
+		if (OMRPORT_INVALID_FD != fd) {
+			unlink(filename);
+			/* Set the file size with ftruncate. */
+			if (OMRPORT_INVALID_FD == ftruncate(fd, byteAmount)) {
+				close(fd);
+				fd = OMRPORT_INVALID_FD;
+			}
+		}
+		if (OMRPORT_INVALID_FD == fd) {
+			/* Revert to using an ANONYMOUS mmap. */
+			useBackingSharedTmpFile = FALSE;
+			flags = 0;
+			useBackingFile = set_flags_for_mmap(&flags);
+			Trc_PRT_vmem_reserve_tempfile_not_created(filename, byteAmount);
+		}
+	}
+	if (useBackingFile) {
 		fd = portLibrary->file_open(portLibrary, "/dev/zero", EsOpenRead | EsOpenWrite, 0);
 	}
 
-	if (((OMRPORT_INVALID_FD == fd) && (!useBackingSharedFile && !useBackingFile)) ||
-	   ((OMRPORT_INVALID_FD != fd) && (useBackingSharedFile || useBackingFile))) {
-
+	BOOLEAN backingFileNeeded = useBackingFile || useBackingSharedFile || useBackingSharedTmpFile;
+	if ((OMRPORT_INVALID_FD != fd) || !backingFileNeeded) {
```

Comment:
This is not an equivalent condition.

Reply:
It is logically equivalent based on the code above which does not allow a valid open `fd` if you don't need one.
The intent is to proceed with the mmap if either
1. You need a backing file and you managed to successfully open one
2. You don't need a backing file. We expect fd to be invalid in this case, so we don't need to test it.

Reply:
The old condition was equivalent to `(OMRPORT_INVALID_FD == fd) != backingFileNeeded`; the new condition may be sufficient, but it's not equivalent. The body executes when `(OMRPORT_INVALID_FD != fd) && !backingFileNeeded` where the old code would not.

Reply:
> The body executes when (OMRPORT_INVALID_FD != fd) && !backingFileNeeded where the old code would not.

My point was that, this scenario cannot happen given the state of the current code. We are protecting against something that cannot happen.
However, I am willing to make it strictly equivalent if you think it's important.

Reply:
I think it's reasonable to simplify the condition, but in the new form I don't think the new variable `backingFileNeeded` is needed, given it's only used once.

Reply:
I have eliminated the new variable in the latest push.